storage/cloudspanner/log_storage.go

```diff
@@ -577,70 +580,85 @@ func (tx *logTX) DequeueLeaves(ctx context.Context, limit int, cutoff time.Time)
-	timeBucket := int64(((now.Unix() + cfg.NumUnseqBuckets/2) % cfg.NumUnseqBuckets) << 8)
+	// Select a prefix that is likley to be on a different span server to spread load.
+	prefix := int64((((now.Unix() + cfg.NumUnseqBuckets/2) % cfg.NumUnseqBuckets) << 8))
 
 	// Choose a starting point in the merkle prefix range, and calculate the
 	// start/limit of the merkle range we'll dequeue from.
 	// It seems to be much better to tune for keeping this range small, and allow
 	// the signer to run multiple times per second than try to dequeue a large batch
 	// which spans a large number of merkle prefixes.
-	merklePrefix := rand.Int63n(256)
-	startBucket := timeBucket | merklePrefix
-	numMerkleBuckets := int64(256 * tx.ls.opts.DequeueAcrossMerkleBucketsRangeFraction)
-	merkleLimit := merklePrefix + numMerkleBuckets
-	if merkleLimit > 0xff {
-		merkleLimit = 0xff
-	}
-	limitBucket := timeBucket | merkleLimit
-
-	stmt := spanner.NewStatement(`
-			SELECT 
-			  Bucket, 
-			  QueueTimestampNanos, 
-			  MerkleLeafHash, 
-			  LeafIdentityHash
-			FROM 
-			  Unsequenced
-			WHERE 
-			  TreeID = @tree_id AND 
-			  Bucket >= @start_bucket AND 
-			  Bucket <= @limit_bucket
-			LIMIT 
-			  @max_num`)
-	stmt.Params["tree_id"] = tx.treeID
-	stmt.Params["start_bucket"] = startBucket
-	stmt.Params["limit_bucket"] = limitBucket
-	stmt.Params["max_num"] = limit
+	suffixStart := rand.Int63n(suffixBuckets)
+	suffixFraction := float64(cfg.NumMerkleBuckets) / float64(suffixBuckets)
+	if tx.ls.opts.DequeueAcrossMerkleBuckets {
+		suffixFraction = tx.ls.opts.DequeueAcrossMerkleBucketsRangeFraction
+	}
+	suffixEnd := suffixStart + int64(suffixBuckets*suffixFraction)
+
+	keysets := []spanner.KeySet{}
+	if suffixEnd <= 0xff {
+		keysets = append(keysets,
+			spanner.KeyRange{
+				Start: spanner.Key{tx.treeID, prefix | suffixStart},
+				End:   spanner.Key{tx.treeID, prefix | suffixEnd},
+				Kind:  spanner.ClosedClosed,
+			})
+	} else {
+		// The range is too big and wraps around, overflowing a byte value, so we'll
+		// start the second range at 0x00 and end at the upper limit modulo 256:
+		suffixEnd %= 0x100
+		keysets = append(keysets,
+			spanner.KeyRange{
+				Start: spanner.Key{tx.treeID, prefix | suffixStart},
+				End:   spanner.Key{tx.treeID, prefix | 0xff},
+				Kind:  spanner.ClosedClosed,
+			},
+			spanner.KeyRange{
+				Start: spanner.Key{tx.treeID, prefix | 0x00},
+				End:   spanner.Key{tx.treeID, prefix | suffixEnd},
+				Kind:  spanner.ClosedClosed,
+			})
+	}
 
 	ret := make([]*trillian.LogLeaf, 0, limit)
-	rows := tx.stx.Query(ctx, stmt)
-	if err := rows.Do(func(r *spanner.Row) error {
-		var l trillian.LogLeaf
-		var qe QueuedEntry
-		if err := r.Columns(&qe.bucket, &qe.timestamp, &l.MerkleLeafHash, &l.LeafIdentityHash); err != nil {
-			return err
-		}
+	rows := tx.stx.Read(ctx, unseqTable, spanner.KeySets(keysets...),
+		[]string{"Bucket", colQueueTimestampNanos, colMerkleLeafHash, colLeafIdentityHash})
+	defer rows.Stop()
+	for {
```

Comment:
Would `rows.Do(...)` make the `switch` unnecessary? It would require a special `errors.New("break")` for stopping after the limit though.

Reply:
Could do that as wel